lib/src/controllers/cartesian_controller.dart

```diff
@@ -160,120 +125,100 @@ class CartesianController extends ChangeNotifier
       results.minYRange = 0.0;
     }
 
-    minXRange = results.minXRange;
-    maxXRange = results.maxXRange;
-    minYRange = results.minYRange;
-    maxYRange = results.maxYRange;
+    return results;
   }
 
   @override
-  void aggregateData(List<CartesianSeries> data) {
-    for (var i = 0; i < data.length; i++) {
-      final series = data[i];
-      series.when(onBarSeries: (barSeries) {
-        // Invalidate Painter for BarSeries
-        if (painters.getOrNull(i).runtimeType != BarPainter) {
-          painters[i] = BarPainter(useGraphUnits: false);
-        } else {
-          // Update if needed.
-        }
-
-        for (var j = 0; j < barSeries.barData.length; j++) {
-          final barGroup = barSeries.barData[j];
-
-          var config = cachedConfigs.getOrNull(barSeries);
-          if (config == null) {
-            config = BarSeriesConfig();
-            cachedConfigs[barSeries] = config;
-          }
-          assert(config is BarSeriesConfig);
-          (config as BarSeriesConfig)
-              .updateEdges(barGroup, _updateMinMaxValues);
-        }
-      });
+  CartesianData constructState(List<CartesianSeries> newData) {
+    // Set our Range Holders
+    var maxXValue = 0.0;
+    var maxYValue = 0.0;
+    var minXValue = 0.0;
+    var minYValue = 0.0;
+    // Set our State holder
+    var states = <PaintingState>[];
+
+    for (var i = 0; i < newData.length; i++) {
+      final series = newData[i];
+      series.when(
+        onBarSeries: (barSeries) {
+          // Invalidate Painter for BarSeries
+          var painter = BarPainter(useGraphUnits: false);
+          var config = BarSeriesConfig();
+
+          config.calcBarDataRange(barSeries.barData, (minX, maxX, minY, maxY) {
+            minXValue = min(minX, minXValue);
+            maxXValue = max(maxX, maxXValue);
+            minYValue = min(minY, minYValue);
+            maxYValue = max(maxY, maxYValue);
+          });
+
+          states.add(
+            BarSeriesState(data: barSeries, config: config, painter: painter),
+          );
+        },
+      );
     }
+
+    // Invalidate the RangeData
+    var results = _invalidateRange(maxXValue, maxYValue, minXValue, minYValue);
+    return CartesianData(state: states, range: results);
   }
 
   @override
-  CartesianConfig? getConfig(CartesianSeries series) => cachedConfigs[series];
+  CartesianData setData(List<CartesianSeries> data) {
+    // Get the cacheKey as a Set of our CartesianSeries.
+    Set<CartesianSeries> cacheKey = Set.from(data);
```

Comment:
What's the advantage of a set over a custom equatable wrapper?

Reply:
Set internally has a methods `setEquals` that compares it's children based on equality. Since all children `CartesianSeries` are Equatable classes, it can do equality checks for all the elements without having the need for a custom wrapper.

Reply:
Changed `Set` to `EquatableList`